src/shogun/kernel/GaussianKernel.cpp

```diff
@@ -82,86 +68,55 @@ CSGObject *CGaussianKernel::shallow_copy() const
 	// with the implement here
 	ASSERT(typeid(*this) == typeid(CGaussianKernel))
 	CGaussianKernel *ker = new CGaussianKernel(cache_size, get_width());
-	if (lhs)
+	if (lhs && rhs)
 	{
 		ker->init(lhs, rhs);
+		ker->m_distance->init(lhs, rhs);
 	}
 	return ker;
 }
 
 void CGaussianKernel::cleanup()
 {
-	if (sq_lhs != sq_rhs)
-		SG_FREE(sq_rhs);
-	sq_rhs = NULL;
-
-	SG_FREE(sq_lhs);
-	sq_lhs = NULL;
-
 	CKernel::cleanup();
-}
-
-void CGaussianKernel::precompute_squared_helper(float64_t* &buf, CDotFeatures* df)
-{
-	ASSERT(df)
-	int32_t num_vec=df->get_num_vectors();
-	buf=SG_MALLOC(float64_t, num_vec);
-
-	for (int32_t i=0; i<num_vec; i++)
-		buf[i]=df->dot(i,df, i);
+	m_distance->cleanup();
 }
 
 bool CGaussianKernel::init(CFeatures* l, CFeatures* r)
 {
-	///free sq_{r,l}hs first
 	cleanup();
-
-	CDotKernel::init(l, r);
-	precompute_squared();
+	CShiftInvariantKernel::init(l, r);
+	m_distance->init(l, r);
 	return init_normalizer();
 }
 
-float64_t CGaussianKernel::compute(int32_t idx_a, int32_t idx_b)
-{
-    float64_t result=distance(idx_a,idx_b);
-    return CMath::exp(-result);
-}
-
-void CGaussianKernel::load_serializable_post() throw (ShogunException)
+void CGaussianKernel::set_width(float64_t w)
 {
-	CKernel::load_serializable_post();
-	precompute_squared();
+	REQUIRE(w>0, "width (%f) must be positive\n",w);
+	m_log_width=CMath::log(w/2.0)/2.0;
 }
 
-void CGaussianKernel::precompute_squared()
+float64_t CGaussianKernel::get_width() const
 {
-	if (!lhs || !rhs)
-		return;
-
-	precompute_squared_helper(sq_lhs, (CDotFeatures*) lhs);
-
-	if (lhs==rhs)
-		sq_rhs=sq_lhs;
-	else
-		precompute_squared_helper(sq_rhs, (CDotFeatures*) rhs);
+	return CMath::exp(m_log_width*2.0)*2.0;
 }
 
-SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(
-		const TParameter* param, index_t index)
+SGMatrix<float64_t> CGaussianKernel::get_parameter_gradient(const TParameter* param, index_t index)
 {
-	REQUIRE(lhs && rhs, "Features not set!\n")
+	REQUIRE(lhs, "Left hand side features must be set!\n")
+	REQUIRE(rhs, "Rightt hand side features must be set!\n")
 
 	if (!strcmp(param->m_name, "log_width"))
 	{
 		SGMatrix<float64_t> derivative=SGMatrix<float64_t>(num_lhs, num_rhs);
-
 		for (int j=0; j<num_lhs; j++)
+		{
 			for (int k=0; k<num_rhs; k++)
 			{
-				float64_t element=distance(j,k);
-				derivative(j,k)=exp(-element)*element*2.0;
+				float64_t element=distance(j, k);
+				derivative(j, k)=exp(-element)*element*2.0;
```

Comment:
for fun and profit: http://stackoverflow.com/questions/10552280/fast-exp-calculation-possible-to-improve-accuracy-without-losing-too-much-perfo/10792513#10792513